File: Fortran/4_datatypes/derived.f90
module print_helpers
    contains

    subroutine print_typeinfo(datatype, typename)
        use mpi
        implicit none
        integer :: datatype
        character*(*) :: typename

        integer ierr, typesize
        integer(KIND=MPI_ADDRESS_KIND) :: lb, extent

        call MPI_TYPE_SIZE(datatype, typesize, ierr)
        call MPI_TYPE_GET_EXTENT(datatype, lb, extent, ierr)

        write (*, fmt="(a,3(a,i0))") typename, ": size=", typesize, " lb=", &
                lb, ", extent=", extent
    end subroutine print_typeinfo

    subroutine print_matrix(matrix, matrix_dim, rank)
        implicit none
        character(len=40) :: fmt
        integer :: matrix_dim, i, rank
        integer, dimension(matrix_dim,matrix_dim) :: matrix

        write (fmt, *) '(i1,a,', matrix_dim, '(" ",i3))'

        do i = 1, matrix_dim
            write (*,fmt) rank, ":", matrix(i,:)
        end do
    end subroutine

end module print_helpers

program datatypes

    use mpi
    use print_helpers
    implicit none

    integer :: i,j
    integer :: ierr, myRank, numProcs, tag = 0
    integer, dimension(MPI_STATUS_SIZE) :: status

    integer :: contigtype, vectortype
    integer :: numblocks, blocklength, stride

    integer, parameter :: matrix_dim = 10
    integer, dimension(matrix_dim, matrix_dim) :: matrix
    character(len=40) :: fmt
    write (fmt, *) '(i1,a,', matrix_dim, '(" ",i3))'

    call MPI_INIT(ierr)

    call MPI_COMM_RANK(MPI_COMM_WORLD, myRank, ierr)
    call MPI_COMM_SIZE(MPI_COMM_WORLD, numProcs, ierr)

    if (myRank .eq. 0) then
        do i = 1, matrix_dim
            do j = 1, matrix_dim
                matrix(j,i) = (j-1)*100 + i
            end do
        end do
    else
        do i = 1, matrix_dim
            do j = 1, matrix_dim
                matrix(j,i) = -1
            end do
        end do
    end if

    if (myRank .eq. 0) then
        numblocks = matrix_dim
        blocklength = 1
        stride = matrix_dim

        ! TODO: Create a vector type for sending columns

        call print_typeinfo(vectortype, "vector")
        call print_matrix(matrix, matrix_dim, myRank);

        ! TODO: Send 10 rows to rank 1

        ! TODO: Free type after use
    else
        ! TODO: Create a vector type for sending columns

        call print_typeinfo(contigtype, "contiguous")

        ! TODO: Receive 10 rows as columns from rank 0

        call print_matrix(matrix, matrix_dim, myRank)

        ! TODO: Free type after use
    endif

    call MPI_FINALIZE(ierr)

end program datatypes

File: C/4_datatypes/derived.c
#include <stdio.h>
#include <stdlib.h>

#include <mpi.h>

const int MATRIX_DIM = 10;

#define INITIALIZE_MATRIX(matrix, size, expr) \
    for (int i = 0; i < size; i++) \
        for( int j = 0; j < size; j++) \
            matrix[i * size + j] = expr;

void print_matrix(int* matrix, int size)
{
    for (int i = 0; i < size; i++)
    {
        for( int j = 0; j < size; j++)
        {
            printf("%3i ", matrix[i*size+j]);
        }
        printf("\n");
    }
}

void print_typeinfo(MPI_Datatype type, const char* typename)
{
    int size;
    MPI_Aint lb, extent;
    MPI_Type_size(type, &size);
    MPI_Type_get_extent(type, &lb, &extent);
    printf("%s: size=%d, lb=%ld, extent=%ld\n", typename, size, lb, extent);
}

int main(int argc, char** argv)
{
    int my_rank, num_ranks, tag = 0;
    MPI_Datatype dtype;

    int* matrix = calloc(MATRIX_DIM*MATRIX_DIM, sizeof(int));

    MPI_Init(&argc, &argv);
    MPI_Comm_rank(MPI_COMM_WORLD, &my_rank);
    MPI_Comm_size(MPI_COMM_WORLD, &num_ranks);

    if (my_rank == 0)
    {
        INITIALIZE_MATRIX(matrix, MATRIX_DIM, i*100+j);
    }
    else
    {
        INITIALIZE_MATRIX(matrix, MATRIX_DIM, 0);
    }

    if (my_rank == 0)
    {
        // TODO: Create a vector type for sending columns
        MPI_Type_vector(MATRIX_DIM, 1, MATRIX_DIM, MPI_INT, &dtype);
        MPI_Type_commit(&dtype);

        print_typeinfo(dtype, "vector");
        print_matrix(matrix, MATRIX_DIM);

        // TODO: Send 10 columns to rank 1
        for (int i = 0; i < 10; i++)
            MPI_Send(&matrix[i], 1, dtype, 1, tag, MPI_COMM_WORLD);    

        // TODO: Free type after use
        MPI_Type_free(&dtype);
    }
    else
    {
        // TODO: Create a contiguous type for receiving columns (as rows)
        MPI_Type_contiguous(MATRIX_DIM, MPI_INT, &dtype);
        MPI_Type_commit(&dtype);

        // TODO: Receive 10 columns as rows from rank 0
        for (int i = 0; i < 10; i++)
            MPI_Recv(&matrix[i

        print_typeinfo(dtype, "contiguous");
        print_matrix(matrix, MATRIX_DIM);

        // TODO: Free type after use
    }

    MPI_Finalize();

    return EXIT_SUCCESS;
}
